src/counter.rs

```diff
@@ -0,0 +1,341 @@
+
+#[derive(Clone, Debug)]
+pub struct Counters {
+    counters: SharedCounters,
+    requests_monitor: RequestsMonitoringCounters,
+}
+
+impl Counters {
+    /// Creates counters from Agreement counter names and starts requests monitoring counters.
+    /// Fails on unsupported counter.
+    pub fn start(counter_names: &Vec<String>) -> anyhow::Result<Self> {
+        let mut counters = Vec::with_capacity(counter_names.len());
+        for counter in counter_names {
+            let counter = SupportedCounter::from_str(counter)?;
+            counters.push(counter);
+        }
+        let counters = Arc::new(RwLock::new(counters));
+
+        let requests_monitor = RequestsMonitoringCounters::start(counters.clone());
+
+        Ok(Self {
+            counters,
+            requests_monitor,
+        })
+    }
+
+    /// Returns usage reported by counters in Agreement specified order.
+    /// None if Agreement had no counter names.
+    pub async fn current_usage(&self) -> Option<Vec<f64>> {
+        let counters = self.counters.read().await;
+        let current_usage = counters.iter().map(Counter::count).collect();
+        Some(current_usage)
+    }
+
+    pub fn requests_monitor(&self) -> impl RequestsMonitor {
+        self.requests_monitor.clone()
+    }
+}
+
+#[derive(Debug)]
+enum SupportedCounter {
+    Duration(DurationCounter),
+    RequestsDuration(RequestsDurationCounter),
+    RequestsCount(RequestsCounter),
+}
+
+impl FromStr for SupportedCounter {
+    type Err = anyhow::Error;
+
+    fn from_str(counter: &str) -> anyhow::Result<Self, Self::Err> {
+        let counter = match counter {
+            "golem.usage.duration_sec" => SupportedCounter::Duration(Default::default()),
+            "golem.usage.gpu-sec" => SupportedCounter::RequestsDuration(Default::default()),
+            "ai-runtime.requests" => SupportedCounter::RequestsCount(Default::default()),
+            _ => bail!("Unsupported counter: {}", counter),
+        };
+        Ok(counter)
+    }
+}
+
+impl SupportedCounter {
+    fn request_monitoring_counter(&mut self) -> Option<&mut dyn RequestMonitoringCounter> {
+        match self {
+            SupportedCounter::Duration(_) => None,
+            SupportedCounter::RequestsDuration(counter) => Some(counter),
+            SupportedCounter::RequestsCount(counter) => Some(counter),
+        }
+    }
+}
+
+impl Counter for SupportedCounter {
+    fn count(&self) -> f64 {
+        let counter: &dyn Counter = match self {
+            SupportedCounter::Duration(counter) => counter,
+            SupportedCounter::RequestsDuration(counter) => counter,
+            SupportedCounter::RequestsCount(counter) => counter,
+        };
+        counter.count()
+    }
+}
+
+trait Counter {
+    fn count(&self) -> f64;
+}
+
+trait RequestMonitoringCounter: Counter {
+    fn on_request(&mut self, request_time: DateTime<Utc>);
+    fn on_response(&mut self, response_time: DateTime<Utc>);
+}
```

Comment:
Why do you need another trait here if you have `ResponseMonitor` and `RequestMonitor` already?

Reply:
`ResponseMonitor` and `RequestMonitor` are splitted only to handle automatic `on_response` on drop of `ResponseMonitor`. Besides these traits do not accept time as a parameter. I pass time parameter because I want to use same time for every counter.